cdap-ui/app/cdap/components/PipelineDetails/PipelineDetailsTopPanel/PipelineDetailsButtons/PipelineStopButton/PipelineStopPopover.js

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+*/
+
+import PropTypes from 'prop-types';
+import React, {Component} from 'react';
+import IconSVG from 'components/IconSVG';
+import Popover from 'components/Popover';
+import Duration from 'components/Duration';
+import moment from 'moment';
+import {Observable} from 'rxjs/Observable';
+import {setStopError} from 'components/PipelineDetails/store/ActionCreator';
+require('./PipelineStopPopover.scss');
+
+export default class PipelineStopPopover extends Component {
+  static propTypes = {
+    runs: PropTypes.array,
+    stopRun: PropTypes.func
+  };
+
+  state = {
+    runsBeingStopped: [],
+    stopAllBtnLoading: false
+  };
+
+  componentWillReceiveProps(nextProps) {
+    let runs = nextProps.runs;
+
+    // Here 'nextProps.runs' refer to running runs, so the runs that were being stopped
+    // won't even show up in the popover anymore, but we should still reset their states
+    if (!runs || !runs.length) {
+      this.setState({
+        runsBeingStopped: [],
+        stopAllBtnLoading: false
+      });
+    } else {
+      let runsBeingStopped = [...this.state.runsBeingStopped];
+      this.state.runsBeingStopped.forEach(runid => {
+        if (runs.indexOf(runid) === -1) {
+          runsBeingStopped.splice(runsBeingStopped.indexOf(runid), 1);
+        }
+      });
+      this.setState({ runsBeingStopped });
+    }
+  }
+
+  stopBtnAndLabel = () => {
+    return (
+      <div className="btn pipeline-action-btn pipeline-stop-btn">
+        <div className="btn-container">
+          <IconSVG name="icon-stop" />
+          <div className="button-label">
+            Stop
+          </div>
+        </div>
+      </div>
+    );
+  };
+
+  stopAllRuns = () => {
+    if (!Array.isArray(this.props.runs) || typeof this.props.stopRun !== 'function') {
+      return;
+    }
+
+    this.setState({ stopAllBtnLoading: true });
+    let stopRunObservables = this.props.runs.map(run => {
+      return this.props.stopRun(run.runid);
+    });
+    Observable.forkJoin(stopRunObservables)
```

Comment:
I believe if you just simple make a call to stop the program instead of doing individual calls for each run, it will still stop all running runs for that program.

Reply:
I looked at https://docs.cask.co/cdap/current/en/reference-manual/http-restful-api/lifecycle.html#H3789, and it says `A program that is in the STOPPED state cannot be stopped. If there are multiple runs of the program in the RUNNING state, this call will stop one of the runs, but not all of the runs.`, so I didn't try that.

Reply:
I see. There's this API as well you can use:
https://docs.cask.co/cdap/current/en/reference-manual/http-restful-api/lifecycle.html#H3672